_source/_posts/2022-04-26-webauthn-java.md

```diff
@@ -0,0 +1,713 @@
+    AppUser user = userRepo.findByUsername(username);
+    List<Authenticator> auth = authRepository.findAllByUser(user);
+    return auth.stream()
+            .map(credential ->
+                PublicKeyCredentialDescriptor.builder()
+                    .id(credential.getCredentialId())
+                    .build())
+            .collect(Collectors.toSet());
+}
+```
+
+The `getUserHandleForUsername` function is called during credential authentication when the user provides a username. The authenticator looks up generated credentials with the help of the user handle byte array, which the server needs to store for this purpose. The demo application uses this function to look up the user handle that's passed into the `navigator.credential.get()` function in the browser.
+
+```java
+@Override
+public Optional<ByteArray> getUserHandleForUsername(String username) {
+    AppUser user = userRepo.findByUsername(username);
+    return Optional.of(user.getHandle());
+}
+```
+
+The `getUsernameForUserHandle` function lets applications handle logins without usernames. The server organizes and recognizes users by means of usernames, but only user handles are necessary for the WebAuthn API to function. `getUsernameForUserHandle` can be used in login flows that don't require usernames up front. Hypothetically, an application could entirely dispense with usernames and identify users exclusively by unique byte array identifiers, though this application does not explore that possibility.
+
+```java
+@Override
+public Optional<String> getUsernameForUserHandle(ByteArray userHandle) {
+    AppUser user = userRepo.findByHandle(userHandle);
+    return Optional.of(user.getUsername());
+}
+```
+
+The `lookup` function is active during the final step of credential verification. A registered authenticator provides an assertion signature and a credential ID to the browser's WebAuthn API, which is sent to the server. The server looks up a stored credential using the credential ID and user handle provided by the authenticator. At this point, the server will use the object returned by this lookup function to validate the assertion signature (and optionally the signature count) from the authenticator.
+
+```java
+@Override
+public Optional<RegisteredCredential> lookup(ByteArray credentialId, ByteArray userHandle) {
+    Optional<Authenticator> auth = authRepository.findByCredentialId(credentialId);
+    return auth.map(credential ->
+        RegisteredCredential.builder()
+            .credentialId(credential.getCredentialId())
+            .userHandle(credential.getUser().getHandle())
+            .publicKeyCose(credential.getPublicKey())
+            .signatureCount(credential.getCount())
+            .build()
+    );
+}
+```
+
+In a similar way, the `lookupAll` function returns a set of `RegisteredCredential` objects. Instead of validating the authenticator's signature, this function ensures that there aren't multiple credentials registered with the same credential ID.
```

Comment:
```suggestion
In a similar way, the `lookupAll()` function returns a set of `RegisteredCredential` objects. Instead of validating the authenticator's signature, this function ensures that there aren't multiple credentials registered with the same credential ID.
```